presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java

```diff
@@ -1342,7 +1342,72 @@ public Node visitNullLiteral(SqlBaseParser.NullLiteralContext context)
     @Override
     public Node visitStringLiteral(SqlBaseParser.StringLiteralContext context)
     {
-        return new StringLiteral(getLocation(context), unquote(context.STRING().getText()));
+        return new StringLiteral(getLocation(context), handleStringExpression(context.stringExpression()));
+    }
+
+    private static String handleStringExpression(SqlBaseParser.StringExpressionContext context)
+    {
+        String rawContent = unquote(context.STRING().getText());
+        if (context.ESCAPE_INTRODUCER() == null) {
+            return rawContent;
+        }
+        String result = "";
+        String currentEscapedCode = "";
+        int lengthBeforeComplete = 0;
+        for (int i = 0; i < rawContent.length(); i++) {
+            if (rawContent.charAt(i) == '\\') {
+                if (lengthBeforeComplete != 0) {
+                    throw new ParsingException(format("Incomplete escaped character: %s", currentEscapedCode), null, context.getStart().getLine(), context.getStart().getCharPositionInLine());
+                }
+                if (i < rawContent.length() - 1 && rawContent.charAt(i + 1) == '+') {
+                    lengthBeforeComplete = 6;
+                    i++;
+                }
+                else {
+                    lengthBeforeComplete = 4;
+                }
+                continue;
+            }
+            if (lengthBeforeComplete != 0) {
+                if (!((rawContent.charAt(i) <= '9' && rawContent.charAt(i) >= '0') ||
+                        (rawContent.charAt(i) <= 'F' && rawContent.charAt(i) >= 'A') ||
+                        (rawContent.charAt(i) <= 'f' && rawContent.charAt(i) >= 'a'))) {
+                    throw new ParsingException(format("Invalid hexadecimal digit: %s", rawContent.charAt(i)), null, context.getStart().getLine(), context.getStart().getCharPositionInLine());
```

Comment:
This line is hard to read. Add a helper (and replace the existing callers in the class):
```java
private static ParsingException parseError(String message, ParserRuleContext context)
{
    return new ParsingException(message, null, context.getStart().getLine(), context.getStart().getCharPositionInLine());
}